src/abstractdataframe/show.jl

```diff
@@ -573,61 +573,186 @@ function _show(io::IO,
+               kwargs...)
+
     _check_consistency(df)
 
-    # we will pass around this buffer to avoid its reallocation in ourstrwidth
-    buffer = IOBuffer(Vector{UInt8}(undef, 80), read=true, write=true)
+    aux = names(df)
+    names_len = textwidth.(aux)
+    maxwidth = max.(9, names_len)
+    names_mat = permutedims(aux)
+    types = eltype.(eachcol(df))
+
+    # NOTE: If we use `type` here, the time to print the first table is 2x more.
+    # This should be something related to type inference.
+    types_str = compacttype.(eltype.(eachcol(df)), maxwidth) |> permutedims
+
+    if allcols && allrows
+        crop = :none
+    elseif allcols
+        crop = :vertical
+    elseif allrows
+        crop = :horizontal
+    else
+        crop = :both
+    end
 
-    nrows = size(df, 1)
-    if rowid !== nothing
-        if size(df, 2) == 0
-            rowid = nothing
-        elseif nrows != 1
-            throw(ArgumentError("rowid may be passed only with a single row data frame"))
+    compact_printing::Bool = get(io, :compact, true)
+
+    # This vector stores the column indices that are only floats. In this case,
+    # the printed numbers will be aligned on the decimal point.
+    float_cols = Int[]
+
+    # These vectors contain the number of the row and the padding that must be
+    # applied so that the float number is aligned with on the decimal point
+    indices = Vector{Int}[]
+    padding = Vector{Int}[]
+
+    # If the screen is limited, we do not need to process all the numbers.
+    dsize = displaysize(io)
+    num_rows, num_cols = size(df)
+
+    if !allcols
+        # Given the spacing, there is no way to fit more than W/9 rows of
+        # floating numbers in the screen, where W is the display width.
+        Δc = clamp(div(dsize[2], 9), 0, num_cols)
+    else
+        Δc = num_cols
+    end
+
+    if !allrows
+        # Get the maximum number of lines that we can display given the screen size.
+        Δr = clamp(dsize[1] - 4, 0, num_rows)
+    else
+        Δr = num_rows
+    end
+
+    Δr_lim = cld(Δr, 2)
+
+    # Do not align the numbers if there are more than 500 rows.
+    if Δr ≤ 500
+        for i = 1:Δc
+            # Analyze the order of the number to compute the maximum padding
+            # that must be applied to align the numbers at the decimal point.
+            if nonmissingtype(types[i]) <: AbstractFloat
+                max_pad_i = 0
+                order_i = zeros(Δr)
+                indices_i = zeros(Δr)
+
+                for k = 1:Δr
+                    # We need to process the top and bottom of the table because
+                    # we are cropping in the middle.
+
+                    kr =  k ≤ Δr_lim ? k : num_rows - (k - Δr_lim) + 1
+
+                    v = df[kr, i]
+
+                    order_v = 0
+
+                    if v isa Number
+                        abs_v = abs(v)
+                        log_v = (!isinf(v) && !isnan(v) && abs_v > 1) ? floor(Int, log10(abs_v))::Int : 0
```

Comment:
`isnan` isn't defined for all `Number` types, so to be safe better check the type too:
```suggestion
                        log_v = (v isa Union{Real, Complex} && !isinf(v) && !isnan(v) && abs_v > 1) ?
                            floor(Int, log10(abs_v))::Int : 0
```